stablehlo/reference/Ops.cpp

```diff
@@ -102,6 +102,123 @@ Tensor evalClampOp(const Tensor &min, const Tensor &operand, const Tensor &max,
   return result;
 }
 
+Tensor evalCompareOp(const Tensor &lhs, const Tensor &rhs,
+                     ComparisonDirection comparisonDirection,
+                     std::optional<ComparisonType> compareType,
+                     TensorType resultType) {
+  Tensor result(resultType);
+  auto elementTy = result.getType().dyn_cast<TensorType>().getElementType();
+  auto comparisonType = ComparisonType::NOTYPE;
+  if (compareType && *compareType != ComparisonType::NOTYPE) {
+    comparisonType = *compareType;
+  } else if (isSupportedUnsignedIntegerType(elementTy) ||
+             isSupportedBooleanType(elementTy)) {
+    comparisonType = ComparisonType::UNSIGNED;
+  } else if (isSupportedSignedIntegerType(elementTy)) {
+    comparisonType = ComparisonType::SIGNED;
+  } else if (isSupportedFloatType(elementTy) ||
+             isSupportedComplexType(elementTy)) {
+    comparisonType = ComparisonType::FLOAT;
+  } else {
+    report_fatal_error(invalidArgument("Unsupported element type: %s",
+                                       debugString(elementTy).c_str()));
+  }
+
+  for (auto it = result.index_begin(); it != result.index_end(); ++it) {
+    bool cmpResult = false;
+    switch (comparisonDirection) {
+      case ComparisonDirection::EQ: {
+        if (comparisonType == ComparisonType::TOTALORDER) {
+          auto lhsFloat = lhs.get(*it).getFloatValue();
+          auto rhsFloat = rhs.get(*it).getFloatValue();
+          cmpResult = lhs.get(*it).getFloatValue().bitcastToAPInt() ==
+                      rhs.get(*it).getFloatValue().bitcastToAPInt();
+        } else {
+          cmpResult = lhs.get(*it) == rhs.get(*it);
+        }
+        break;
+      }
+      case ComparisonDirection::NE: {
+        if (comparisonType == ComparisonType::TOTALORDER) {
+          auto lhsFloat = lhs.get(*it).getFloatValue();
+          auto rhsFloat = rhs.get(*it).getFloatValue();
+          cmpResult = lhs.get(*it).getFloatValue().bitcastToAPInt() !=
+                      rhs.get(*it).getFloatValue().bitcastToAPInt();
+        } else {
+          cmpResult = lhs.get(*it) != rhs.get(*it);
+        }
+        break;
+      }
+      case ComparisonDirection::GE: {
+        if (comparisonType == ComparisonType::TOTALORDER) {
+          auto lhsFloat = lhs.get(*it).getFloatValue();
+          auto rhsFloat = rhs.get(*it).getFloatValue();
+          if (lhsFloat.isNegative() ^ rhsFloat.isNegative())
+            cmpResult = rhsFloat.isNegative();
+          else
+            cmpResult =
+                lhsFloat.isNegative()
+                    ? lhsFloat.bitcastToAPInt().ule(rhsFloat.bitcastToAPInt())
+                    : lhsFloat.bitcastToAPInt().uge(rhsFloat.bitcastToAPInt());
```

Comment:
Good question! I was also weary of using it. IEEE section 5.10 c.3 says:
```
If x and y represent the same floating-point datum:

i) If x and y have negative sign,
totalOrder(x, y) is true if and only if the exponent of x ≥ the exponent of y

ii) otherwise,
totalOrder(x, y) is true if and only if the exponent of x ≤ the exponent of
```
This behavior can be captured by interpreting the bits as an integer and comparing it like normal integer comparison.